components/style/style_adjuster.rs

```diff
@@ -85,6 +90,26 @@ impl<'a> StyleAdjuster<'a> {
         }
     }
 
+    /// Change writing mode of text frame for text-combine-upright.
+    /// It is safe to look at the parent's style because we are looking at
+    /// inherited properties, and ::-moz-text never matches any rules.
+    #[cfg(feature = "gecko")]
+    fn adjust_for_text_combine_upright(&mut self,
+                                       layout_parent_style: &ComputedValues) {
+        if let Some(p) = self.pseudo {
+            if *p.as_atom() == atom!(":-moz-text") {
+                use computed_values::text_combine_upright::T as text_combine_upright;
+                use computed_values::writing_mode::T as writing_mode;
+                let parent_writing_mode = layout_parent_style.get_inheritedbox().clone_writing_mode();
+                let parent_text_combine_upright = layout_parent_style.get_inheritedtext().clone_text_combine_upright();
+                if parent_writing_mode != writing_mode::horizontal_tb &&
+                   parent_text_combine_upright == text_combine_upright::all {
+                    self.style.mutate_inheritedbox().set_writing_mode(writing_mode::horizontal_tb);
```

Comment:
How does this work, exactly?

In particular, we don't traverse text in the traversal at all...